src/core/lib/event_engine/ares_resolver.cc

```diff
@@ -0,0 +1,714 @@
+      });
+}
+
+void AresResolver::OnReadable(FdNode* fd_node, absl::Status status) {
+  grpc_core::MutexLock lock(&mutex_);
+  GPR_ASSERT(fd_node->readable_registered);
+  fd_node->readable_registered = false;
+  GRPC_ARES_RESOLVER_TRACE_LOG("OnReadable: fd: %d; request: %p; status: %s",
+                               fd_node->as, this, status.ToString().c_str());
+  if (status.ok() && !shutting_down_) {
+    ares_process_fd(channel_, fd_node->as, ARES_SOCKET_BAD);
+  } else {
+    // If error is not absl::OkStatus() or the resolution was cancelled, it
+    // means the fd has been shutdown or timed out. The pending lookups made
+    // on this request will be cancelled by the following ares_cancel(). The
+    // remaining file descriptors in this request will be cleaned up in the
+    // following Work() method.
+    ares_cancel(channel_);
+  }
+  CheckSocketsLocked();
+}
+
+void AresResolver::OnWritable(FdNode* fd_node, absl::Status status) {
+  grpc_core::MutexLock lock(&mutex_);
+  GPR_ASSERT(fd_node->writable_registered);
+  fd_node->writable_registered = false;
+  GRPC_ARES_RESOLVER_TRACE_LOG("OnWritable: fd: %d; request:%p; status: %s",
+                               fd_node->as, this, status.ToString().c_str());
+  if (status.ok() && !shutting_down_) {
+    ares_process_fd(channel_, ARES_SOCKET_BAD, fd_node->as);
+  } else {
+    // If error is not absl::OkStatus() or the resolution was cancelled, it
+    // means the fd has been shutdown or timed out. The pending lookups made
+    // on this request will be cancelled by the following ares_cancel(). The
+    // remaining file descriptors in this request will be cleaned up in the
+    // following Work() method.
+    ares_cancel(channel_);
+  }
+  CheckSocketsLocked();
+}
+
+// In case of non-responsive DNS servers, dropped packets, etc., c-ares has
+// intelligent timeout and retry logic, which we can take advantage of by
+// polling ares_process_fd on time intervals. Overall, the c-ares library is
+// meant to be called into and given a chance to proceed name resolution:
+//   a) when fd events happen
+//   b) when some time has passed without fd events having happened
+// For the latter, we use this backup poller. Also see
+// https://github.com/grpc/grpc/pull/17688 description for more details.
+void AresResolver::OnAresBackupPollAlarm() {
+  grpc_core::MutexLock lock(&mutex_);
+  GRPC_ARES_RESOLVER_TRACE_LOG(
+      "request:%p OnAresBackupPollAlarm shutting_down=%d.", this,
+      shutting_down_);
+  if (!shutting_down_) {
+    for (const auto& fd_node : fd_node_list_) {
+      if (!fd_node->already_shutdown) {
+        GRPC_ARES_RESOLVER_TRACE_LOG(
+            "request:%p OnAresBackupPollAlarm; ares_process_fd. fd=%s", this,
+            fd_node->polled_fd->GetName());
+        ares_socket_t as = fd_node->polled_fd->GetWrappedAresSocketLocked();
+        ares_process_fd(channel_, as, as);
+      }
+    }
+    EventEngine::Duration next_ares_backup_poll_alarm_duration =
```

Comment:
Suggest calling `MaybeStartTimerLocked()` instead of duplicating its code here.

Reply:
Good point! Done.